src/crypto/ed448_suite.c

```diff
@@ -0,0 +1,205 @@
+            data_to_verify->length, prehashed, context, context_length) == DECAF_SUCCESS);
+
+}
+
+wickr_buffer_t *ed448_dh_derive_public_key(const wickr_buffer_t *private_key_data)
+{
+    if (!private_key_data || private_key_data->length != DH_448_PRIVATE_KEY_LENGTH)
+        return NULL;
+
+    wickr_buffer_t *public_key_data = wickr_buffer_create_empty_zero(DH_448_PUBLIC_KEY_LENGTH);
+
+    if (!public_key_data)
+        return NULL;
+
+    decaf_x448_derive_public_key(public_key_data->bytes, private_key_data->bytes);
+    // This function does not allow for failure
+
+    return public_key_data;  
+}
+
+wickr_buffer_t *ed448_dh_shared_secret(const wickr_ecdh_params_t *params)
+{
+    if (!params)
+        return NULL;
+    
+    const wickr_ec_key_t *local_key_pair = params->local_key;
+    const wickr_ec_key_t *peer_public_key = params->peer_key;
+
+    if (!local_key_pair || !peer_public_key)
+        return NULL;
+
+    if (local_key_pair->curve.identifier != EC_CURVE_ID_ED448_GOLDILOCKS ||
+        peer_public_key->curve.identifier != EC_CURVE_ID_ED448_GOLDILOCKS)
+        return NULL;
+        
+    const wickr_buffer_t *local_private_key_data = local_key_pair->pri_data;
+    const wickr_buffer_t *peer_public_key_data = peer_public_key->pub_data;
+
+    if (!local_private_key_data || !peer_public_key_data)
+        return NULL;
+
+    if (local_private_key_data->length != DH_448_PRIVATE_KEY_LENGTH ||
+        peer_public_key_data->length != DH_448_PUBLIC_KEY_LENGTH)
+        return NULL;
+
+    wickr_buffer_t *shared_secret = wickr_buffer_create_empty_zero(DH_448_SHARED_SECRET_LENGTH);
+
+    /* Compute the shared secret */
+    decaf_error_t result = decaf_x448(shared_secret->bytes, peer_public_key_data->bytes,
+                                      local_private_key_data->bytes);
+    
+    if (result == DECAF_FAILURE) {
+        wickr_buffer_destroy(&shared_secret);
+        return NULL;
+    }
+
+    /* Run the ECDH shared secret through HKDF with the provided salt and info from the ECDH params */
+    wickr_kdf_result_t *kdf_result = wickr_perform_kdf_meta(params->kdf_info, shared_secret);
+    wickr_buffer_t *final_buffer = wickr_buffer_copy(kdf_result->hash);
+    wickr_kdf_result_destroy(&kdf_result);
+    
+    return final_buffer;
+}
+
+#define hash_ctx_t   decaf_shake256_ctx_t
+#define hash_init    decaf_shake256_init
+#define hash_update  decaf_shake256_update
+#define hash_final   decaf_shake256_final
+#define hash_destroy decaf_shake256_destroy
+#define hash_hash    decaf_shake256_hash
+
+wickr_buffer_t *ed448_shake256_raw(const wickr_buffer_t * data, uint16_t output_length)
+{
+    if (!data)
+        return NULL;
+    
+    hash_ctx_t hash;
+    hash_init(hash);
+    hash_update(hash,data->bytes,data->length);
+   
+    wickr_buffer_t * result = wickr_buffer_create_empty_zero(output_length);
+
+    hash_final(hash,result->bytes,result->length);
```

Comment:
can this fail?

Reply:
No, this one cannot. But hash_update might, and that needed some fixes. Thanks!